server/tokenlimiter.go

```diff
@@ -13,29 +13,43 @@
 
 package server
 
+import "sync/atomic"
+
 // Token is used as a permission to keep on running.
 type Token struct {
 }
 
 // TokenLimiter is used to limit the number of concurrent tasks.
 type TokenLimiter struct {
-	count uint
-	ch    chan *Token
+	capacity  uint32
+	allocated uint32
+	ch        chan *Token
 }
 
 // Put releases the token.
 func (tl *TokenLimiter) Put(tk *Token) {
 	tl.ch <- tk
+	atomic.AddUint32(&tl.allocated, ^uint32(0))
```

Comment:
Uber's atomic wrapper supports `sub` semantic, which may be better? https://github.com/uber-go/atomic#usage

Reply:
OK.